### PR TITLE
Add custom analysis related APIs

### PR DESCRIPTION
**Pull request by bhufmann**
**Changes (multiple commits)**
- Support parentId and creation configuration in output descriptors
- Add APIs to support configurable output (data provider)
  - Add configuration service per output (data provider)
  This addition enables clients to create derived data providers from
  an existing data provider.
  - Add unit tests
  - Add cli to invoke new endpoints
- Support JSON file for load-configuration and update-configuration
  This allows users to use a file with JSON parameters.
- Extract method for parsing/validating options.params/option.json_file

**To Test**

With Trace Compass Trace Server of today, the new CLI commands below can be used to configure InAndOut analysis, with LTTng Kernel traces or UST traces with function entry or exit.

```
  ./tsp_cli_client --list-output-configuration-sources OUTPUT_ID --uuid UUID
  ./tsp_cli_client --list-output-configuration-source TYPE_ID --output-id OUTPUT_ID --uuid UUID
  ./tsp_cli_client --create-output OUTPUT_ID --uuid UUID --json-file absolute-path-to-json-file
  ./tsp_cli_client --delete-output DERIVED_OUTPUT_ID --output-id  OUTPUT_ID --uuid UUID

where:
OUTPUT_ID = org.eclipse.tracecompass.incubator.inandout.core.analysis.inAndOutdataProviderFactory
TYPE_ID = org.eclipse.tracecompass.incubator.internal.inandout.core.config
UUID = UUID of experiment
```

Content of JSON file:
```
{
   "name": "my other InAndOut",
   "description" : "my description",
   "sourceTypeId": "org.eclipse.tracecompass.incubator.internal.inandout.core.config",
   "parameters": {
       "specifiers" : [
           {
           "label":"latency",
           "inRegex":"(\\S*)_entry",
           "outRegex":"(\\S*)_exit",
           "contextInRegex":"(\\S*)_entry",
           "contextOutRegex":"(\\S*)_exit",
           "classifier":"CPU"
           }
       ]
   }
}

```

**Note:** This PR includes changes of #86 which needs to merged before.

Signed-off-by: Bernd Hufmann <bernd.hufmann@ericsson.com>

### CODE REVIEW

File: tsp/output_descriptor.py [hunk 1]
```diff
@@ -126,16 +144,24 @@ def to_json(self):
...
-                TYPE_KEY: obj.type,
-                # Hide non-TSP fields
-                # QUERY_PARAMETERS_KEY: obj.query_parameters,
-                # START_TIME_KEY: obj.start,
-                # END_TIME_KEY: obj.end,
-                # IS_FINAL_KEY: obj.final,
-                # COMPATIBLE_PROVIDERS_KEY: obj.compatible_providers
-            }
+            result = {}
+            # optinal parent_id
```
Comment (PatrickTasse) [inline tsp/output_descriptor.py:148]: optional

File: tsp_cli_client [hunk 2]
```diff
@@ -41,6 +41,7 @@ from tsp.tsp_client import TspClient
 
 TRACE_MISSING = "Trace UUID is missing"
 
+
```
Comment (PatrickTasse) [inline tsp_cli_client:45]: extra line or intentional?

edit: Ok, It gets removed in the next commit.

File: README.md [hunk 3]
```diff
@@ -160,6 +160,18 @@ optional arguments:
   --params PARAMS       comma separated key value pairs (key1=val1,key2=val2)
   --get-health          Get the health status of the server
   --get-identifier      Identify important information regarding the server and the system
+  --list-output-configuration-sources OUTPUT_ID
+                        Get available configuration sources for a given experiment and output
+  --list-output-configuration-source TYPE_ID
```
Comment (PatrickTasse) [inline README.md:165]: Is it missing OUTPUT_ID ?

Oh, I see it has to be provided as a separate argument. It's not really clear which arguments must be provided together. But that's the way it is I guess. The user will get a proper error message when missing.

File: README.md [hunk 4]
```diff
@@ -160,6 +160,18 @@ optional arguments:
   --params PARAMS       comma separated key value pairs (key1=val1,key2=val2)
   --get-health          Get the health status of the server
   --get-identifier      Identify important information regarding the server and the system
+  --list-output-configuration-sources OUTPUT_ID
```
Comment (PatrickTasse) [inline README.md:163]: Do these new options need to be added after line 91 also?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (bhufmann) [reply README.md:163]: Yes, they are missing.

File: README.md (hunk 3 above)
Comment (bhufmann) [reply README.md:165]: Yes, it's not clear at all. That's why I added the examples below to show all options needed per command. The help output (`./tsp_typescript_client --help`) is not good.

File: tsp/output_descriptor.py (hunk 1 above)
Comment (bhufmann) [reply tsp/output_descriptor.py:148]: Done

File: tsp_cli_client (hunk 2 above)
Comment (bhufmann) [reply tsp_cli_client:45]: yes

**Review by PatrickTasse — DISMISSED** [review]

**Review by PatrickTasse — APPROVED** [review]

**Pull request merged**